mongoengine/document.py

```diff
@@ -488,7 +488,13 @@ def reload(self):
         """
         id_field = self._meta['id_field']
         collection = self._get_collection()
-        son = collection.find_one(self._db_object_key)
+        # If this is a lazy object, we only have the ID field and don't want to
+        # call _db_object_key, since _db_object_key could fetch (reload) the
+        # object.
+        if self._lazy:
+            son = collection.find_one({ '_id': self.pk })
+        else:
+            son = collection.find_one(self._db_object_key)
```

Comment:
What happens if we have a doc with a shard key on multiple fields, and you fetch a doc with `Doc.objects.only('id').first()` and then call `doc.reload()`. Will it incorrectly call `find_one` with `{ other_fields: None }`?